File: tools/automation/src/ultra205-defaults-evidence.ts
import { createHash } from "node:crypto";
import { chmod, mkdir, readFile, readdir, stat, writeFile } from "node:fs/promises";
import path from "node:path";

import {
  internalCommandSpec,
  type AutomationCategory,
  type Ultra205DefaultsEvidence,
} from "./contracts.generated.js";
import type { ProcessPort } from "./process.js";
import {
  captureSystemInfoEvidence,
  SystemInfoEvidenceError,
  type SystemInfoEvidenceOptions,
} from "./system-info-evidence.js";
import type { WebSocketFactory } from "./websocket.js";
import { assertWithinWorkspace } from "./workspace.js";

export type Ultra205DefaultsEvidenceOptions = SystemInfoEvidenceOptions;

type JsonObject = Readonly<Record<string, unknown>>;
type FailureCategory = Extract<AutomationCategory, "hardware_blocked" | "evidence_invalid" | "timeout" | "process_failed">;

const configuredDefaultFieldCount = 27;
const apiVisibleDefaultFieldCount = 23;
const retainedAttestation = "ultra205_config_defaults schema_version=1 matching_fields=27 total_fields=27 all_match=true mineonboot_disabled=true redacted=true";

export class Ultra205DefaultsEvidenceError extends Error {
  public constructor(
    public readonly category: FailureCategory,
    message: string,
    public readonly publicValue: Readonly<Record<string, unknown>>,
  ) {
    super(message);
    this.name = "Ultra205DefaultsEvidenceError";
  }
}

function failure(category: FailureCategory, message: string) {
  return new Ultra205DefaultsEvidenceError(category, message, {
    stage: "ultra205_defaults_capture",
    recovery_complete: true,
    recovery_flash_used: false,
    secondary_recovery_failure: false,
  });
}

function sha256(value: string | Buffer): string {
  return createHash("sha256").update(value).digest("hex");
}

function object(value: unknown, context: string): JsonObject {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw failure("evidence_invalid", `${context} must be an object`);
  }
  return value as JsonObject;
}

function parseSeedFixture(document: string): ReadonlyMap<string, string> {
  const values = new Map<string, string>();
  for (const line of document.split(/\r?\n/u)) {
    if (line === "" || line.startsWith("#") || line === "key,type,encoding,value") continue;
    const fields = line.split(",");
    if (fields.length !== 4) throw failure("evidence_invalid", "Ultra 205 seed fixture row is invalid");
    const [key, type, encoding, value] = fields;
    if (key === undefined || type === undefined || encoding === undefined || value === undefined) {
      throw failure("evidence_invalid", "Ultra 205 seed fixture row is incomplete");
    }
    if (type === "namespace") continue;
    if (type !== "data" || !new Set(["string", "u16"]).has(encoding) || values.has(key)) {
      throw failure("evidence_invalid", "Ultra 205 seed fixture identity is invalid");
    }
    values.set(key, value);
  }
  if (values.size !== configuredDefaultFieldCount + 2) {
    throw failure("evidence_invalid", "Ultra 205 seed fixture field count is invalid");
  }
  return values;
}

function fixtureValue(values: ReadonlyMap<string, string>, key: string): string {
  const maybeValue = values.get(key);
  if (maybeValue === undefined) throw failure("evidence_invalid", "Ultra 205 seed fixture is incomplete");
  return maybeValue;
}

function integerFixture(values: ReadonlyMap<string, string>, key: string): number {
  const parsed = Number(fixtureValue(values, key));
  if (!Number.isSafeInteger(parsed)) throw failure("evidence_invalid", "Ultra 205 seed fixture integer is invalid");
  return parsed;
}

function apiDefaultsMatch(snapshot: JsonObject, values: ReadonlyMap<string, string>): boolean {
  const expected: ReadonlyArray<readonly [string, unknown]> = [
    ["hostname", fixtureValue(values, "hostname")],
    ["stratumURL", fixtureValue(values, "stratumurl")],
    ["stratumPort", integerFixture(values, "stratumport")],
    ["stratumTLS", integerFixture(values, "stratumtls")],
    ["stratumCert", fixtureValue(values, "stratumcert")],
    ["stratumUser", fixtureValue(values, "stratumuser")],
    ["stratumSuggestedDifficulty", integerFixture(values, "stratumdiff")],
    ["stratumExtranonceSubscribe", integerFixture(values, "stratumxnsub") !== 0],
    ["fallbackStratumURL", fixtureValue(values, "fbstratumurl")],
    ["fallbackStratumPort", integerFixture(values, "fbstratumport")],
    ["fallbackStratumTLS", integerFixture(values, "fbstratumtls")],
    ["fallbackStratumCert", fixtureValue(values, "fbstratumcert")],
    ["fallbackStratumUser", fixtureValue(values, "fbstratumuser")],
    ["fallbackStratumSuggestedDifficulty", integerFixture(values, "fbstratumdiff")],
    ["fallbackStratumExtranonceSubscribe", integerFixture(values, "fbstratumxnsum") !== 0],
    ["frequency", integerFixture(values, "asicfrequency")],
    ["coreVoltage", integerFixture(values, "asicvoltage")],
    ["ASICModel", fixtureValue(values, "asicmodel")],
    ["boardVersion", fixtureValue(values, "boardversion")],
    ["rotation", integerFixture(values, "rotation")],
    ["autofanspeed", integerFixture(values, "autofanspeed")],
    ["manualFanSpeed", integerFixture(values, "fanspeed")],
    ["overheat_mode", integerFixture(values, "overheat_mode")],
  ];
  return expected.length === apiVisibleDefaultFieldCount
    && expected.every(([field, expectedValue]) => snapshot[field] === expectedValue)
    && snapshot["startMiningOnBoot"] === false
    && snapshot["miningPaused"] === true;
}

async function privateModesValid(root: string): Promise<boolean> {
  if (((await stat(root)).mode & 0o777) !== 0o700) return false;
  for (const entry of await readdir(root, { withFileTypes: true })) {
    const child = path.join(root, entry.name);
    if (entry.isDirectory()) {
      if (!await privateModesValid(child)) return false;
    } else if (entry.isFile() && ((await stat(child)).mode & 0o777) !== 0o600) {
      return false;
    }
  }
  return true;
}

async function writePrivateJson(output: string, value: unknown): Promise<void> {
  await writeFile(output, `${JSON.stringify(value, null, 2)}\n`, {
    encoding: "utf8", flag: "wx", mode: 0o600,
  });
  await chmod(output, 0o600);
}

export async function captureUltra205DefaultsEvidence(
  workspaceRoot: string,
  options: Ultra205DefaultsEvidenceOptions,
  processPort: ProcessPort,
  flashProgram: string,
  systemInfoValidatorProgram: string,
  defaultsValidatorProgram: string,
  maybeWebSocketFactory?: WebSocketFactory,
): Promise<Ultra205DefaultsEvidence> {
  const privateRoot = assertWithinWorkspace(workspaceRoot, options.privateRoot);
  const projection = assertWithinWorkspace(workspaceRoot, options.projection);
  const fixturePath = path.join(workspaceRoot, "crates/bitaxe-config/fixtures/ultra-205-defaults.csv");
  const fixtureDocument = await readFile(fixturePath, "utf8");
  const fixture = parseSeedFixture(fixtureDocument);
  const systemProjection = path.join(privateRoot, "system-info-projection.private.json");
  let systemInfo;
  try {
    systemInfo = await captureSystemInfoEvidence(
      workspaceRoot,
      { ...options, projection: systemProjection },
      processPort,
      flashProgram,
      systemInfoValidatorProgram,
      maybeWebSocketFactory,
    );
  } catch (error) {
    if (error instanceof SystemInfoEvidenceError) {
      throw new Ultra205DefaultsEvidenceError(error.category, error.message, error.publicValue);
    }
    throw error;
  }
  await chmod(systemProjection, 0o600);

  const api = object(JSON.parse(await readFile(path.join(privateRoot, "api.private.json"), "utf8")), "HTTP defaults snapshot");
  const websocketEnvelope = object(JSON.parse(await readFile(path.join(privateRoot, "websocket.private.json"), "utf8")), "WebSocket defaults envelope");
  const websocket = object(websocketEnvelope["data"], "WebSocket defaults snapshot");
  const retained = await readFile(path.join(privateRoot, "retained-log.private.txt"), "utf8");
  if (!apiDefaultsMatch(api, fixture) || !apiDefaultsMatch(websocket, fixture)) {
    throw failure("evidence_invalid", "API-visible Ultra 205 defaults do not match");
  }
  if (!new Set(retained.split(/\r?\n/u)).has(retainedAttestation)) {
    throw failure("evidence_invalid", "retained Ultra 205 defaults attestation is missing");
  }

  const manifestDocument = await readFile(assertWithinWorkspace(workspaceRoot, options.packageManifest), "utf8");
  const systemInfoDocument = await readFile(systemProjection, "utf8");
  const evidence: Ultra205DefaultsEvidence = {
    schema_version: "bitaxe-ultra205-defaults-evidence-v1",
    board: 205,
    source_commit: systemInfo.source_commit,
    reference_commit: systemInfo.reference_commit,
    package_manifest_sha256: systemInfo.package_manifest_sha256,
    workflow: {
      schema_version: "bitaxe-workflow-identity-v1",
      command: "capture-ultra205-defaults-evidence",
      request_sha256: sha256(JSON.stringify({
        manifest: sha256(manifestDocument),
        fixture: sha256(fixtureDocument),
        system_info: sha256(systemInfoDocument),
        timeout: options.captureTimeoutSeconds,
      })),
    },
    detector_admitted: true,
    boot_observed: true,
    system_info: systemInfo,
    defaults: {
      configured_default_field_count: configuredDefaultFieldCount,
      firmware_matching_field_count: configuredDefaultFieldCount,
      firmware_all_defaults_match: true,
      api_visible_default_field_count: apiVisibleDefaultFieldCount,
      http_defaults_match: true,
      websocket_defaults_match: true,
      retained_attestation_matches: true,
      mining_on_boot_disabled: true,
      exact_seed_fixture_sha256: sha256(fixtureDocument),
      system_info_evidence_sha256: sha256(systemInfoDocument),
    },
    mining_state: "disabled",
    hardware_control_state: "disabled",
    cleanup_complete: true,
    private_modes_valid: true,
    redaction_status: "passed",
  };
  const candidate = path.join(privateRoot, "ultra205-defaults-evidence.private.json");
  await writePrivateJson(candidate, evidence);
  if (!await privateModesValid(privateRoot)) {
    throw failure("evidence_invalid", "private Ultra 205 defaults artifacts have invalid modes");
  }
  let validation;
  try {
    validation = await processPort.run(internalCommandSpec(defaultsValidatorProgram, [candidate], (value) => value));
  } catch {
    throw failure("process_failed", "Ultra 205 defaults validator launch failed");
  }
  if (validation.timedOut) throw failure("timeout", "Ultra 205 defaults validation timed out");
  if (validation.exitCode !== 0) throw failure("evidence_invalid", "Ultra 205 defaults validation failed");
  await mkdir(path.dirname(projection), { recursive: true });
  await writeFile(projection, `${JSON.stringify(evidence, null, 2)}\n`, { encoding: "utf8", flag: "wx" });
  return evidence;
}

File: crates/bitaxe-automation-contracts/src/ultra205_defaults_evidence.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{
    AutomationCommand, SystemInfoEvidence, WorkflowIdentity, ULTRA205_DEFAULTS_EVIDENCE_SCHEMA,
};

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct Ultra205DefaultsObservationEvidence {
    pub configured_default_field_count: u16,
    pub firmware_matching_field_count: u16,
    pub firmware_all_defaults_match: bool,
    pub api_visible_default_field_count: u16,
    pub http_defaults_match: bool,
    pub websocket_defaults_match: bool,
    pub retained_attestation_matches: bool,
    pub mining_on_boot_disabled: bool,
    pub exact_seed_fixture_sha256: String,
    pub system_info_evidence_sha256: String,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct Ultra205DefaultsEvidence {
    pub schema_version: String,
    pub board: u16,
    pub source_commit: String,
    pub reference_commit: String,
    pub package_manifest_sha256: String,
    pub workflow: WorkflowIdentity,
    pub detector_admitted: bool,
    pub boot_observed: bool,
    pub system_info: SystemInfoEvidence,
    pub defaults: Ultra205DefaultsObservationEvidence,
    pub mining_state: String,
    pub hardware_control_state: String,
    pub cleanup_complete: bool,
    pub private_modes_valid: bool,
    pub redaction_status: String,
}

impl Ultra205DefaultsEvidence {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.schema_version != ULTRA205_DEFAULTS_EVIDENCE_SCHEMA || self.board != 205 {
            return Err("Ultra 205 defaults evidence schema or board is invalid");
        }
        if self.workflow.schema_version != "bitaxe-workflow-identity-v1"
            || self.workflow.command != AutomationCommand::CaptureUltra205DefaultsEvidence
        {
            return Err("Ultra 205 defaults workflow identity is invalid");
        }
        self.system_info.validate()?;
        if self.source_commit != self.system_info.source_commit
            || self.reference_commit != self.system_info.reference_commit
            || self.package_manifest_sha256 != self.system_info.package_manifest_sha256
        {
            return Err("Ultra 205 defaults system info identity is inconsistent");
        }
        for digest in [
            self.package_manifest_sha256.as_str(),
            self.workflow.request_sha256.as_str(),
            self.defaults.exact_seed_fixture_sha256.as_str(),
            self.defaults.system_info_evidence_sha256.as_str(),
        ] {
            if digest.len() != 64
                || !digest
                    .bytes()
                    .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
            {
                return Err("Ultra 205 defaults evidence digest is invalid");
            }
        }
        let defaults = &self.defaults;
        if defaults.configured_default_field_count != 27
            || defaults.firmware_matching_field_count != 27
            || !defaults.firmware_all_defaults_match
            || defaults.api_visible_default_field_count != 23
            || !defaults.http_defaults_match
            || !defaults.websocket_defaults_match
            || !defaults.retained_attestation_matches
            || !defaults.mining_on_boot_disabled
        {
            return Err("Ultra 205 configured defaults are incomplete");
        }
        if !self.detector_admitted
            || !self.boot_observed
            || self.mining_state != "disabled"
            || self.hardware_control_state != "disabled"
            || !self.cleanup_complete
            || !self.private_modes_valid
            || self.redaction_status != "passed"
        {
            return Err("Ultra 205 defaults safety or privacy evidence is invalid");
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        AutomationCommand, SystemInfoEvidence, SystemInfoObservationEvidence, WorkflowIdentity,
    };

    use super::{Ultra205DefaultsEvidence, Ultra205DefaultsObservationEvidence};

    fn workflow(command: AutomationCommand, value: char) -> WorkflowIdentity {
        WorkflowIdentity {
            schema_version: "bitaxe-workflow-identity-v1".to_owned(),
            command,
            request_sha256: value.to_string().repeat(64),
        }
    }

    fn system_info() -> SystemInfoEvidence {
        SystemInfoEvidence {
            schema_version: "bitaxe-system-info-evidence-v1".to_owned(),
            board: 205,
            source_commit: "a".repeat(40),
            reference_commit: "b".repeat(40),
            package_manifest_sha256: "c".repeat(64),
            workflow: workflow(AutomationCommand::CaptureSystemInfoEvidence, 'd'),
            detector_admitted: true,
            boot_observed: true,
            same_origin_observed: true,
            system_info: SystemInfoObservationEvidence {
                boot_session_sha256: "e".repeat(64),
                http_revision: 7,
                websocket_revision: 8,
                same_boot_session: true,
                websocket_revision_not_earlier: true,
                field_contract_schema: "bitaxe-system-info-field-contract-v1".to_owned(),
                field_contract_sha256: "f".repeat(64),
                required_field_count: 94,
                unconditional_field_count: 87,
                conditional_field_count: 7,
                http_unconditional_fields_complete: true,
                websocket_unconditional_fields_complete: true,
                http_field_types_match: true,
                websocket_field_types_match: true,
                inactive_block_fields_absent: true,
                confirmed_setting_fields_present: true,
                retained_http_tuple_matches: true,
                retained_websocket_tuple_matches: true,
            },
            mining_state: "disabled".to_owned(),
            hardware_control_state: "disabled".to_owned(),
            cleanup_complete: true,
            redaction_status: "passed".to_owned(),
        }
    }

    fn valid_evidence() -> Ultra205DefaultsEvidence {
        Ultra205DefaultsEvidence {
            schema_version: "bitaxe-ultra205-defaults-evidence-v1".to_owned(),
            board: 205,
            source_commit: "a".repeat(40),
            reference_commit: "b".repeat(40),
            package_manifest_sha256: "c".repeat(64),
            workflow: workflow(AutomationCommand::CaptureUltra205DefaultsEvidence, '1'),
            detector_admitted: true,
            boot_observed: true,
            system_info: system_info(),
            defaults: Ultra205DefaultsObservationEvidence {
                configured_default_field_count: 27,
                firmware_matching_field_count: 27,
                firmware_all_defaults_match: true,
                api_visible_default_field_count: 23,
                http_defaults_match: true,
                websocket_defaults_match: true,
                retained_attestation_matches: true,
                mining_on_boot_disabled: true,
                exact_seed_fixture_sha256: "2".repeat(64),
                system_info_evidence_sha256: "3".repeat(64),
            },
            mining_state: "disabled".to_owned(),
            hardware_control_state: "disabled".to_owned(),
            cleanup_complete: true,
            private_modes_valid: true,
            redaction_status: "passed".to_owned(),
        }
    }

    #[test]
    fn complete_closed_projection_is_accepted() {
        // Arrange
        let evidence = valid_evidence();

        // Act
        let result = evidence.validate();

        // Assert
        assert_eq!(result, Ok(()));
    }

    #[test]
    fn partial_firmware_match_is_rejected() {
        // Arrange
        let mut evidence = valid_evidence();
        evidence.defaults.firmware_matching_field_count = 26;

        // Act
        let result = evidence.validate();

        // Assert
        assert_eq!(result, Err("Ultra 205 configured defaults are incomplete"));
    }
}
